lib/easypost/exceptions/api/api_error.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require_relative '../easy_post_error'
+require 'easypost/constants'
+
+class EasyPost::Exceptions::ApiError < EasyPost::Exceptions::EasyPostError
+  attr_reader :status_code, :code, :errors
+
+  def initialize(message, status_code = nil, error_code = nil, sub_errors = nil)
+    super message
+    @status_code = status_code
+    @code = error_code
+    @errors = sub_errors
+  end
+
+  def pretty_print
+    error_string = "#{code} (#{status_code}): #{message}"
+    errors&.each do |error|
+      error_string += "\nField: #{error.field}\nMessage: #{error.message}"
+    end
+    error_string
+  end
+
+  # Recursively traverses a JSON element to extract error messages and returns them as a comma-separated string.
+  def self.collect_error_messages(error_message, messages_list)
+    case error_message
+    when Hash
+      error_message.each_value { |value| collect_error_messages(value, messages_list) }
+    when Array
+      error_message.each { |value| collect_error_messages(value, messages_list) }
+    else
+      messages_list.push(error_message.to_s)
+    end
+
+    messages_list.join(', ')
+  end
+
+  def self.handle_api_error(response)
+    status_code = response.code
+    status_code = status_code.to_i if status_code.is_a?(String)
+
+    if status_code >= 200 && status_code <= 299
+      return nil
+    end
+
+    # Try to parse the response body as JSON
+    begin
+      error_data = JSON.parse(response.body)['error']
+
+      error_message = error_data['message']
+      error_type = error_data['code']
+      errors = error_data['errors']&.map do |error|
+        EasyPost::Models::Error.from_api_error_response(error)
+      end
+    rescue StandardError
+      error_message = response.code.to_s
+      error_type = EasyPost::Constants::API_ERROR_DETAILS_PARSING_ERROR
+      errors = nil
+    end
+
+    cls = exception_cls_from_status_code(status_code)
+
+    if cls == EasyPost::Exceptions::UnknownHttpError
+      error_message = EasyPost::Constants::UNEXPECTED_HTTP_STATUS_CODE % status_code
+      raise EasyPost::Exceptions::UnknownHttpError.new(error_message, status_code, nil, nil)
+    end
+
+    # Return (don't throw here) an instance of the appropriate error class
+    cls.new(error_message, status_code, error_type, errors)
+  end
+
+  def self.exception_cls_from_status_code(status_code)
+    # rubocop:disable Lint/DuplicateBranch
+    case status_code
+    when 0
+      EasyPost::Exceptions::ConnectionError
+    when 100, 101, 102, 103
```

Comment:
Question: do we want to throw an error when the HTTP is 1xx? I don't think we have done that in other libs

Reply:
We do in .NET. But also, our API should NEVER throw a 1XX.

Reply:
I suggest to remove them, it will fall to default clause `UnknownError` anyway

Reply:
I'll second this suggestion, it's cleaner since we don't need to throw something different than the default

Reply:
This has been removed